src/libponyrt/actor/actor.c

```diff
@@ -19,31 +19,148 @@
+// in fact muted.
+// 2. It's ok for a muted actor to be seen as unmuted in a transient fashion
+// across actors
+//
+// If rule #1 is violated, we might end up deadlocking because an actor was
+// muted for sending to an actor that might never be unmuted (because it isn't
+// muted). The actor muted actor would continue to remain muted and the actor
+// incorrectly seen as muted became actually muted and then unmuted.
+//
+// If rule #2 is violated, then a muted actor will receive from 1 to a few
+// additional messages and the sender won't be muted. As this is a transient
+// situtation that should be shortly rectified, there's no harm done.
+//
+// Our handling of atomic operations in `mute_actor`
+// and `unmute_actor` are to assure that both rules aren't violated.
+
+static void mute_actor(pony_actor_t* actor)
+{
+  set_sync_flag(actor, SYNC_FLAG_MUTED);
+  DTRACE1(ACTOR_MUTED, (uintptr_t)actor);
+}
+
+void ponyint_unmute_actor(pony_actor_t* actor)
+{
+  unset_sync_flag(actor, SYNC_FLAG_MUTED);
+  DTRACE1(ACTOR_UNMUTED, (uintptr_t)actor);
+}
 
-bool has_flag(pony_actor_t* actor, uint8_t flag)
+static bool triggers_muting(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  return (flags & flag) != 0;
+  return has_sync_flag_any(actor, SYNC_FLAG_OVERLOADED |
+    SYNC_FLAG_UNDER_PRESSURE | SYNC_FLAG_MUTED);
 }
 
-static void set_flag(pony_actor_t* actor, uint8_t flag)
+static void actor_setoverloaded(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  atomic_store_explicit(&actor->flags, flags | flag, memory_order_relaxed);
+  pony_assert(!ponyint_is_cycle(actor));
+  set_sync_flag(actor, SYNC_FLAG_OVERLOADED);
+  DTRACE1(ACTOR_OVERLOADED, (uintptr_t)actor);
 }
 
-static void unset_flag(pony_actor_t* actor, uint8_t flag)
+static void actor_unsetoverloaded(pony_actor_t* actor)
 {
-  uint8_t flags = atomic_load_explicit(&actor->flags, memory_order_relaxed);
-  atomic_store_explicit(&actor->flags, flags & (uint8_t)~flag,
-    memory_order_relaxed);
+  pony_ctx_t* ctx = pony_ctx();
+  unset_sync_flag(actor, SYNC_FLAG_OVERLOADED);
+  DTRACE1(ACTOR_OVERLOADED_CLEARED, (uintptr_t)actor);
+  if (!has_sync_flag(actor, SYNC_FLAG_UNDER_PRESSURE))
+  {
+    ponyint_sched_start_global_unmute(ctx->scheduler->index, actor);
+  }
+}
+
+static void maybe_mute(pony_ctx_t* ctx, pony_actor_t* to)
+{
+  if(ctx->current != NULL)
+  {
+    // only mute a sender IF:
+    // 1. the receiver is overloaded/under pressure/muted
+    // AND
+    // 2. the sender isn't overloaded or under pressure
+    // AND
+    // 3. we are sending to another actor (as compared to sending to self)
+    if(triggers_muting(to) &&
+      !has_sync_flag_any(ctx->current, SYNC_FLAG_OVERLOADED |
+        SYNC_FLAG_UNDER_PRESSURE) &&
+      ctx->current != to)
+    {
+      ponyint_sched_mute(ctx, ctx->current, to);
+    }
+  }
```

Comment:
this logic is rather different compared to the previous, did you pull a lot of additional logic into here?

Reply:
this looks very different. than what you are replacing. this looks like a mistake to me.